src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,413 @@
+                });
+                var results = await this.leaseProvider.Acquire(LeaseCategory, leaseRequests.ToArray());
+                //add successfully acquired queue to myQueues list
+                for (int i = 0; i < results.Length; i++)
+                {
+                    if (results[i].StatusCode == ResponseCode.OK)
+                    {
+                        this.myQueues.Add(new AcquiredQueue(expectedQueues[i], results[i].AcquiredLease));
+                    }
+                }
+                //if reached expectedTotalLeaseCount
+                if (this.myQueues.Count >= expectedTotalLeaseCount)
+                {
+                    break;
+                }
+            }
+
+            this.logger.Info($"AcquireLeasesToMeetExpectation: finished. Now own {this.myQueues.Count} queues. Used attemps : {attempts}, Current minimumReponsibility : {this.minimumResponsibilty}, current maximumResponsibility : {this.maximumRespobsibility}");
+        }
+        
+        private async Task RenewLeases()
+        {
+            var results = await this.leaseProvider.Renew(LeaseCategory, this.myQueues.Select(queue => queue.AcquiredLease).ToArray());
+            var updatedQueues = new List<AcquiredQueue>();
+            //update myQueues list with successfully renewed leases
+            for (int i = 0; i < results.Count(); i++)
+            {
+                var result = results[i];
+                if (result.StatusCode == ResponseCode.OK)
+                {
+                    updatedQueues.Add(new AcquiredQueue(this.myQueues[i].QueueId, result.AcquiredLease));
+                }
+            }
+            this.myQueues.Clear();
+            this.myQueues = updatedQueues;
+            this.logger.Info($"RenewLeases: finished, currently own queues : {this.myQueues.Count}, current minimumResponsibilty : {this.minimumResponsibilty}, current maximunResponsibilty : {this.maximumRespobsibility}");
+        }
+
+        private void CalculateResponsibility()
+        {
+            int activeBuckets = 0;
+            if (isStarting)
+            {
+                activeBuckets = this.deploymentConfig.GetAllSiloNames().Count;
+            }
+            else
+            {
+                activeBuckets = QueueBalancerUtilities.GetActiveSilos(this.siloStatusOracle, this.immatureSilos).Count;
+            }
+            this.minimumResponsibilty = this.allQueues.Count / activeBuckets;
+            //if allQueues count is divisible by active bukets, then every bucket should take the same count of queues, otherwise, there should be one bucket take 1 more queue
+            if (this.allQueues.Count % activeBuckets == 0)
+                this.maximumRespobsibility = this.minimumResponsibilty;
+            else this.maximumRespobsibility = this.minimumResponsibilty + 1;
+        }
+
+        /// <inheritdoc/>
+        public void SiloStatusChangeNotification(SiloAddress updatedSilo, SiloStatus status)
```

Comment:
I kept this in order to only use the algorithm in GetActiveSilos, which I thought is the optimal/old way to do it. But you mentioned there's a bug there. So I think we can talk about that next week....

the treading issue we talked about is both timer thread and siloStatusOracleNotification thread can be calling MaintainAndBalanceQueues, hence the threading issue.... I changed siloStatusOracleNotification to not trigger `MaintainAndBalanceQueues` anymore... only timer will trigger `MaintainAndBalanceQueues`..... 

Anyway, I kept SiloStatusChangeNotification only because the old GetActiveSilo algorithm.... if that is a incorrect algorithm which has a bug... then no need to keep it.